x-pack/plugin/data-frame/src/main/java/org/elasticsearch/xpack/dataframe/transforms/DataFrameTransformTask.java

```diff
@@ -285,37 +338,174 @@ void persistStateToClusterState(DataFrameTransformState state,
+        private DataFrameTransformProgress progress;
+
+        ClientDataFrameIndexer build(DataFrameTransformTask parentTask) {
+            return new ClientDataFrameIndexer(this.transformId,
+                this.transformsConfigManager,
+                this.transformsCheckpointService,
+                new AtomicReference<>(this.indexerState),
+                this.initialPosition,
+                this.client,
+                this.auditor,
+                this.initialStats,
+                this.transformConfig,
+                this.fieldMappings,
+                this.progress,
+                parentTask);
+        }
+
+        ClientDataFrameIndexerBuilder setClient(Client client) {
+            this.client = client;
+            return this;
+        }
+
+        ClientDataFrameIndexerBuilder setTransformsConfigManager(DataFrameTransformsConfigManager transformsConfigManager) {
+            this.transformsConfigManager = transformsConfigManager;
+            return this;
+        }
+
+        ClientDataFrameIndexerBuilder setTransformsCheckpointService(DataFrameTransformsCheckpointService transformsCheckpointService) {
+            this.transformsCheckpointService = transformsCheckpointService;
+            return this;
+        }
+
+        ClientDataFrameIndexerBuilder setTransformId(String transformId) {
+            this.transformId = transformId;
+            return this;
+        }
 
+        ClientDataFrameIndexerBuilder setAuditor(Auditor<DataFrameAuditMessage> auditor) {
+            this.auditor = auditor;
+            return this;
+        }
+
+        ClientDataFrameIndexerBuilder setFieldMappings(Map<String, String> fieldMappings) {
+            this.fieldMappings = fieldMappings;
+            return this;
+        }
+
+        ClientDataFrameIndexerBuilder setTransformConfig(DataFrameTransformConfig transformConfig) {
+            this.transformConfig = transformConfig;
+            return this;
+        }
+
+        DataFrameTransformConfig getTransformConfig() {
+            return this.transformConfig;
+        }
+
+        ClientDataFrameIndexerBuilder setInitialStats(DataFrameIndexerTransformStats initialStats) {
+            this.initialStats = initialStats;
+            return this;
+        }
+
+        ClientDataFrameIndexerBuilder setIndexerState(IndexerState indexerState) {
+            this.indexerState = indexerState;
+            return this;
+        }
+
+        ClientDataFrameIndexerBuilder setInitialPosition(Map<String, Object> initialPosition) {
+            this.initialPosition = initialPosition;
+            return this;
+        }
+
+        Map<String, Object> getInitialPosition() {
+            return this.initialPosition;
+        }
+
+        ClientDataFrameIndexerBuilder setProgress(DataFrameTransformProgress progress) {
+            this.progress = progress;
+            return this;
+        }
+    }
+
+    static class ClientDataFrameIndexer extends DataFrameIndexer {
```

Comment:
I moved this to a static class as it seems to me that it should have been one from the get go. Conflating where the task's responsibilities ends and the indexer's starts has been a continuous problem with this design. This is just a first step in trying to make sure that there is at least some separation of concerns between the two.